### DISCUSSION TITLE
How does autogluon timeseries handle new item_ids

### DISCUSSION THREAD

**Discussion by yvpathirana** [autogluon/autogluon#3382] [Q&A]
I am curious about how the autogluon handles the new item ids which are not in the training set but are in the test set usually most of the other models will raise an error because the item ids are not defined in the training set and most of them models out there does not know how to handle this kind of event.  I  have go through the code but couldn't find it. if anyone can help me with my curiosity I will be very grateful . if anyone knows where code lies please put the link to the relevant Python file.Thank you for taking your time

**Comment by shchur** [comment]
Hi, thanks for the question! As of v0.8, AutoGluon only uses the `item_id` column to differentiate between individual time series in the DataFrame, but does not use the actual value of the `item_id`. Therefore, TimeSeriesPredictor can handle following scenarios:
- the same item_id refers to two completely different time series in `train_data` passed to `fit()` and `data` passed to `predict()`
- `data` passed to `predict()` contains item_ids not seen during training

Note that this might change in a future release - we will mention this in release notes, should this occur.

-----
If you use the `item_id` as a categorical `static_feature`, then values not seen during training will be replaced by the most frequent value of the static feature encountered in the training set (https://github.com/autogluon/autogluon/blob/master/timeseries/src/autogluon/timeseries/utils/features.py#L37).

**Reply by yvpathirana** [reply]
Hi,
thank you for your reply , sorry if I have misunderstood but,
when fitting ARIMA model in Autogluon I think it will fit models for individual time series, then how does new item id will be handled? because for that new item_id Autogluon doesn't have a particular individual model to predict that item. what you are referring is correct for global models. ETS AutoArima AutoETS I suppose these models are not global

**Reply by shchur** [reply]
Thanks, now I understand the question. 

For local models, we do not store the fitted model as these can often use lots of memory (e.g., a fitted ARIMA model from `statsmodels` may sometimes take >10 GB of RAM) and, as you pointed out, require tracking `item_id`, which can be cumbersome. Instead, we delegate all the work for local models to the `predict` method: inside `predict` the model trains on the data and returns the forecast DataFrame.
- Here is the base class for all local models that implements this logic: https://github.com/autogluon/autogluon/blob/master/timeseries/src/autogluon/timeseries/models/local/abstract_local_model.py#L117
- Here is a concrete example for `StatsForecast` models https://github.com/autogluon/autogluon/blob/master/timeseries/src/autogluon/timeseries/models/local/statsforecast.py#L24

Note that, by default, we also cache the predictions of all models, so repeated calls to `predictor.predict()`, `predictor.evaluate()` or `predictor.leaderboard()` will completely skip the fit&predict step for local models.

**Reply by yvpathirana** [reply]
ohh this means we do not need to train much time on local models. anyway in the prediction it will be trained again. Is this correct?
when it training it will know which model is best when we haven't specified a model in predict

**Reply by shchur** [reply]
Exactly, the "fit time" is usually close to zero for local models because actual fitting is done inside the `predict` method.

When you call `TimeSeriesPredictor.fit()`, we will generate predictions with local model to score them (= determine if they should be included in the ensemble), but when you make predictions on new data, the models will be re-trained from scratch.

**Reply by hengamirah** [reply]
Hi, I have 10 different items that requires forecasting and a model for each own items,
does '.fit' fits the data with item_id provide a "generalized" model based on all 10 items? or does it create a model for each item_id during the 'predict' method?

**Reply by shchur** [reply]
Hi @hengamirah, this depends on the model. For local models like AutoETS or AutoARIMA, a separate model is created on the fly for each time series. For global models like DeepAR or TFT, a single model is trained that "works well" on average for all the time series. See Section 2.1 of https://arxiv.org/pdf/2004.10240 for more details on the distinction.
